File: python-service/app/adapters/mock_usps.py
"""Mock USPS adapter for development and testing.

This module provides a mock implementation of the USPS adapter interface
that returns deterministic responses without calling the real USPS API.
"""

from app.adapters.usps_adapter import USPSAdapterInterface
from app.models.address import StructuredAddress
from app.models.validation import ValidationResult, ZipcodeCityResult

# Known test addresses that return deterministic standardized responses
_KNOWN_ADDRESSES = {
    "1600 PENNSYLVANIA AVE NW": StructuredAddress(
        street_line_1="1600 PENNSYLVANIA AVE NW",
        street_line_2=None,
        city="WASHINGTON",
        state="DC",
        zipcode="20500",
    ),
}


class MockUSPSAdapter(USPSAdapterInterface):
    """Mock USPS adapter that returns deterministic responses."""

    async def validate_address(self, address: StructuredAddress) -> ValidationResult:
        """Validate a structured address using mock logic.

        Returns invalid status when street_line_1 contains the substring "INVALID".
        Returns a deterministic standardized address for known test addresses.
        Returns valid with uppercased standardized address for all other inputs.
        """
        # Check for "INVALID" substring in street_line_1 (case-sensitive on the keyword)
        if "INVALID" in address.street_line_1:
            return ValidationResult(
                original_address=address,
                standardized_address=None,
                status="invalid",
                error_message="Address not found in USPS database",
            )

        # Check for known test addresses (match on uppercased street_line_1)
        known = _KNOWN_ADDRESSES.get(address.street_line_1.upper())
        if known:
            return ValidationResult(
                original_address=address,
                standardized_address=known,
                status="valid",
            )

        # Default: return valid with uppercased standardized address
        standardized = StructuredAddress(
            street_line_1=address.street_line_1.upper(),
            street_line_2=address.street_line_2.upper() if address.street_line_2 else None,
            city=address.city.upper(),
            state=address.state.upper(),
            zipcode=address.zipcode,
        )
        return ValidationResult(
            original_address=address,
            standardized_address=standardized,
            status="valid",
        )

    async def verify_zipcode_city(self, zipcode: str, city: str) -> ZipcodeCityResult:
        """Verify that a city matches a given zipcode using mock data.

        Returns match for known pairs, mismatch with valid_cities otherwise.
        """
        # Known zipcode-city pairs for mock testing
        known_pairs = {
            "20500": ["WASHINGTON"],
            "10001": ["NEW YORK"],
            "90210": ["BEVERLY HILLS"],
            "60601": ["CHICAGO"],
        }

        zip5 = zipcode[:5]
        valid_cities = known_pairs.get(zip5)

        if valid_cities and city.upper() in valid_cities:
            return ZipcodeCityResult(
                zipcode=zipcode,
                city=city.upper(),
                status="match",
            )

        return ZipcodeCityResult(
            zipcode=zipcode,
            city=city,
            status="mismatch",
            valid_cities=valid_cities or ["UNKNOWN"],
        )

File: express-service/src/adapters/mockUsps.ts
import { StructuredAddress } from '../models/address';
import { ValidationResult, ZipcodeCityResult } from '../models/validation';
import { USPSAdapter } from './uspsAdapter';
import seedData from '../data/seed-addresses.json';

/**
 * Valid US state codes for validation.
 */
const VALID_STATES = new Set([
  'AL','AK','AZ','AR','CA','CO','CT','DE','DC','FL','GA','HI','ID','IL','IN',
  'IA','KS','KY','LA','ME','MD','MA','MI','MN','MS','MO','MT','NE','NV','NH',
  'NJ','NM','NY','NC','ND','OH','OK','OR','PA','RI','SC','SD','TN','TX','UT',
  'VT','VA','WA','WV','WI','WY'
]);

/**
 * Known zipcode-city pairs for mock verification.
 */
const ZIPCODE_CITY_MAP: Record<string, string[]> = {
  '20500': ['WASHINGTON'],
  '10001': ['NEW YORK'],
  '10005': ['NEW YORK'],
  '10007': ['NEW YORK'],
  '10118': ['NEW YORK'],
  '90210': ['BEVERLY HILLS'],
  '60601': ['CHICAGO'],
  '60606': ['CHICAGO'],
  '95014': ['CUPERTINO'],
  '94025': ['MENLO PARK'],
  '98109': ['SEATTLE'],
  '98052': ['REDMOND'],
  '78701': ['AUSTIN'],
  '94304': ['PALO ALTO'],
  '91608': ['UNIVERSAL CITY'],
  '92802': ['ANAHEIM'],
  '02139': ['CAMBRIDGE'],
  '48105': ['ANN ARBOR'],
  '94105': ['SAN FRANCISCO'],
  '94103': ['SAN FRANCISCO'],
  '33612': ['TAMPA'],
  '33613': ['TAMPA'],
  '62701': ['SPRINGFIELD'],
};

/**
 * Build a lookup map from seed data for known addresses.
 */
const SEED_LOOKUP = new Map<string, typeof seedData[0]>();
for (const entry of seedData) {
  const key = entry.street_line_1.toUpperCase().trim();
  SEED_LOOKUP.set(key, entry);
}

/**
 * Mock USPS adapter using seed data for realistic responses.
 */
export class MockUSPSAdapter implements USPSAdapter {

  async validateAddress(address: StructuredAddress): Promise<ValidationResult> {
    // Check for INVALID keyword
    if (address.street_line_1.toUpperCase().includes('INVALID')) {
      return {
        original_address: address,
        status: 'invalid',
        error_message: 'Address not found in USPS database',
      };
    }

    // Check for invalid state code
    if (!VALID_STATES.has(address.state.toUpperCase())) {
      return {
        original_address: address,
        status: 'invalid',
        error_message: `Invalid state code: ${address.state}`,
      };
    }

    // Look up in seed data
    const key = address.street_line_1.toUpperCase().trim();
    const seedEntry = SEED_LOOKUP.get(key);

    if (seedEntry && seedEntry.valid && seedEntry.standardized) {
      return {
        original_address: address,
        standardized_address: seedEntry.standardized as StructuredAddress,
        status: 'valid',
      };
    }

    if (seedEntry && !seedEntry.valid) {
      return {
        original_address: address,
        status: 'invalid',
        error_message: seedEntry.error_message || 'Address not found in USPS database',
      };
    }

    // Default: standardize (uppercase) any address with valid state
    const standardized: StructuredAddress = {
      street_line_1: address.street_line_1.toUpperCase(),
      ...(address.street_line_2 ? { street_line_2: address.street_line_2.toUpperCase() } : {}),
      city: address.city.toUpperCase(),
      state: address.state.toUpperCase(),
      zipcode: address.zipcode,
    };

    return {
      original_address: address,
      standardized_address: standardized,
      status: 'valid',
    };
  }

  async verifyZipcodeCity(zipcode: string, city: string): Promise<ZipcodeCityResult> {
    const zip5 = zipcode.substring(0, 5);
    const validCities = ZIPCODE_CITY_MAP[zip5];

    if (validCities && validCities.includes(city.toUpperCase())) {
      return {
        zipcode,
        city: city.toUpperCase(),
        status: 'match',
      };
    }

    if (validCities) {
      return {
        zipcode,
        city,
        status: 'mismatch',
        valid_cities: validCities,
      };
    }

    // Unknown zipcode
    return {
      zipcode,
      city,
      status: 'mismatch',
      valid_cities: ['UNKNOWN'],
    };
  }
}
